iotfunctions/anomaly.py

```diff
@@ -1285,52 +1662,207 @@ def build_ui(cls):
                 ))
         return (inputs, outputs)
 
 
-class FFTbasedGeneralizedAnomalyScore(GeneralizedAnomalyScore):
+class GeneralizedAnomalyScorev2(Standard_Scaler):
     """
-    An unsupervised and robust anomaly detection function.
-     Extracts temporal features from time series data using Fast Fourier Transforms.
-     Applies the GeneralizedAnomalyScore to the features to detect outliers.
-     Moves a sliding window across the data signal and applies the anomaly models to each window.
+    An unsupervised anomaly detection function.
+     Applies the Minimum Covariance Determinant (FastMCD) technique to detect outliers.
+     Moves a sliding window across the data signal and applies the anomaly model to each window.
      The window size is typically set to 12 data points.
-     Try several anomaly detectors on your data and use the one that best fits your data.
+     The normalize switch allows to learn and apply a standard scaler prior to computing the anomaly score.
+     Try several anomaly detectors on your data and use the one that fits your data best.
     """
+    # class variables
+    eval_metric = staticmethod(metrics.r2_score)
 
-    def __init__(self, input_item, windowsize, output_item):
-        super().__init__(input_item, windowsize, output_item)
+    train_if_no_model = True
 
-        self.whoami = 'FFT'
-        self.normalizer = FFT_normalizer
+    def __init__(self, input_item, windowsize, normalize, output_item, expr=None):
+        super().__init__(features=[input_item], targets=[output_item], predictions=None)
 
-        logger.debug('FFT')
+
+        logger.debug(input_item)
+        # do not run score and call transform instead of predict
+
+        self.input_item = input_item
+
+        # use 12 by default
+        self.windowsize, windowoverlap = set_window_size_and_overlap(windowsize)
+
+        # step
+        self.step = self.windowsize - windowoverlap
+
+        self.normalize = normalize
+
+        # assume 1 per sec for now
+        self.frame_rate = 1
+
+        self.dampening = 1  # dampening - dampen anomaly score
+
+        self.output_item = output_item
+
+        self.normalizer = Generalized_normalizer
+
+        self.whoami = 'GAMv2'
+
+
+    def prepare_data(self, dfEntity):
+
+        logger.debug(self.whoami + ': prepare Data')
+
+        # operate on simple timestamp index
+        if len(dfEntity.index.names) > 1:
+            index_names = dfEntity.index.names
+            dfe = dfEntity.reset_index().set_index(index_names[0])
+        else:
+            index_names = None
+            dfe = dfEntity
+
+        # interpolate gaps - data imputation
+        try:
+            dfe = dfe.interpolate(method="time")
+        except Exception as e:
+            logger.error('Prepare data error: ' + str(e))
+
+        # one dimensional time series - named temperature for catchyness
+        temperature = dfe[[self.input_item]].fillna(0).to_numpy().reshape(-1,)
```

Comment:
same comment as the one in `KMeansAnomalyScorev2.prepare_data`
`self.input_item` is not the scaled_input, we are running the anomaly detection on the original data and not the scaled data

Reply:
Same again.
prepare_data is now moved into Standard_Scaler and returns the predicted/scaled column for all subclasses.